File: lib/py/hld.py
import sys
sys.setrecursionlimit(10**6)

class HLD:
    def __init__(self, n, adj, root=1):
        self.n = n
        self.adj = adj
        self.parent = [0] * (n + 1)
        self.depth = [0] * (n + 1)
        self.heavy = [0] * (n + 1)
        self.head = [0] * (n + 1)
        self.pos = [0] * (n + 1)
        self.cur_pos = 0
        
        self.parent[root] = 0
        self.depth[root] = 0
        self._dfs(root)
        self._decompose(root, root)
        
    def _dfs(self, v):
        size = 1
        max_c_size = 0
        for c in self.adj[v]:
            if c != self.parent[v]:
                self.parent[c] = v
                self.depth[c] = self.depth[v] + 1
                c_size = self._dfs(c)
                size += c_size
                if c_size > max_c_size:
                    max_c_size = c_size
                    self.heavy[v] = c
        return size
    
    def _decompose(self, v, h):
        self.head[v] = h
        self.pos[v] = self.cur_pos
        self.cur_pos += 1
        
        if self.heavy[v] != 0:
            self._decompose(self.heavy[v], h)
            
        for c in self.adj[v]:
            if c != self.parent[v] and c != self.heavy[v]:
                self._decompose(c, c)
                
    # 경로 쿼리 템플릿
    # def query(self, u, v):
    #     res = 0
    #     while self.head[u] != self.head[v]:
    #         if self.depth[self.head[u]] > self.depth[self.head[v]]:
    #             u, v = v, u
    #         # res += seg.query(self.pos[self.head[v]], self.pos[v])
    #         v = self.parent[self.head[v]]
    #     if self.depth[u] > self.depth[v]:
    #         u, v = v, u
    #     # res += seg.query(self.pos[u], self.pos[v])
    #     return res

File: lib/cpp/hld.cpp
#include <algorithm>
#include <iostream>
#include <vector>


using namespace std;

// Heavy-Light Decomposition
const int MAX = 100005;
vector<int> adj[MAX];
int parent[MAX], depth[MAX], heavy[MAX], head[MAX], pos[MAX];
int cur_pos;

// 1. DFS to find heavy edges and subtree sizes
int dfs(int v) {
  int size = 1;
  int max_c_size = 0;
  for (int c : adj[v]) {
    if (c != parent[v]) {
      parent[c] = v;
      depth[c] = depth[v] + 1;
      int c_size = dfs(c);
      size += c_size;
      if (c_size > max_c_size) {
        max_c_size = c_size;
        heavy[v] = c;
      }
    }
  }
  return size;
}

// 2. Decompose chains
void decompose(int v, int h) {
  head[v] = h;
  pos[v] = cur_pos++;
  if (heavy[v] != -1)
    decompose(heavy[v], h);
  for (int c : adj[v]) {
    if (c != parent[v] && c != heavy[v])
      decompose(c, c);
  }
}

// 초기화
void init_hld(int root) {
  fill(heavy, heavy + MAX, -1);
  parent[root] = -1;
  depth[root] = 0;
  cur_pos = 0;
  dfs(root);
  decompose(root, root);
}

// 경로 쿼리 예시 (Segment Tree 필요)
// int query(int a, int b) {
//     int res = 0;
//     for (; head[a] != head[b]; b = parent[head[b]]) {
//         if (depth[head[a]] > depth[head[b]]) swap(a, b);
//         // res += seg.query(pos[head[b]], pos[b]);
//     }
//     if (depth[a] > depth[b]) swap(a, b);
//     // res += seg.query(pos[a], pos[b]);
//     return res;
// }
